### ISSUE TITLE
Authentication flow help

### ISSUE THREAD

**Issue by KyRooZz** [ronzeidman/ng2-ui-auth#40]
Hi guys,

This is not an issue but I really need help understanding how everything works with Github or any other provider..

Actually I installed the plugin and I did as mentionned in the README, so here is my code so far, which is only client-side : 

import { Component, AfterContentInit } from '@angular/core';
import { ControlGroup, FormBuilder, Validators } from '@angular/common';
import { Router, ROUTER_DIRECTIVES } from '@angular/router';

import { Auth, NG2_UI_AUTH_PROVIDERS } from 'ng2-ui-auth';
import { AuthService } from '../../services/auth.service';
import { FormStateModel } from '../../models/form/form-state.model';
import { LoginRequestModel } from '../../models/auth/login-request.model';
import { FormInputTextComponent } from '../+shared/index';
import { FocusableDirective } from '../../directives/focusable.directive';

const DEFAULT_POST_HEADER: { [name: string]: string } = {
    'Content-Type': 'application/json'
};
const GITHUB_CLIENT_ID = '##MY_GITHUB_ID##';

@Component({
    selector: 'login',
    templateUrl: 'src/components/login/login.component.html',
    providers: [
        NG2_UI_AUTH_PROVIDERS({
            defaultHeaders: DEFAULT_POST_HEADER,
            providers: {
                github: { clientId: GITHUB_CLIENT_ID }
            }
        })],
    directives: [ROUTER_DIRECTIVES, FormInputTextComponent, FocusableDirective],
    pipes: []
})

export class LoginComponent implements AfterContentInit {
    formState: FormStateModel;
    loginForm: ControlGroup;

    private _loginRequest: LoginRequestModel;

    constructor(private auth: Auth, private _formBuilder: FormBuilder, private _router: Router, private _authService: AuthService) {
        if (this._authService.isUserAuthorized()) {
            this._router.navigate(['/dashboard']);
        }

        this.formState = new FormStateModel();

        this.loginForm = this._formBuilder.group({
            'username': ['', Validators.required],
            'password': ['', Validators.required]
        });

        this.loginForm.valueChanges.subscribe(() => {
            this.formState.submitError = false;
        });
    }

    goToMain() {
        this._router.navigate(['/dashboard']);
    }

    authenticate(provider: string) {
        this.auth.authenticate(provider)
            .subscribe(() => () => this.goToMain(),
            (errorMessage: string) => {
                this.formState.submitting = false;
                this.formState.submitError = true;
                this.formState.submitErrorMessage = errorMessage;
            });
    }
}

Github popup is shown but when I click on **authorize application**, I'm getting an error on my login page : 

'SyntaxError: Unexpected token < in JSON at position 0'

So my questions are : 
- Why am i getting this error?
- Do I have to implement a server-side and why, is it not possible to have only a client-side authentication ? Because I just want the user email from the provider and play with it after.

Thanks !

**Comment by ronzeidman** [comment]
Hi,
Usually in angular "SyntaxError: Unexpected token <" actually means "Error 404" this is since every route that does not resolve in an actual file is redirected to the angular's index.html (which obviously starts with <html...). so what the error means that there is some address in your site which should have resolved in a file that doesn't exist, probably due to not implementing the server-side.

As for the second question oauth requires server-side since in order to gain access to your app you have to be able to access the user's private information (like name and email) that is provided by the authorization server (in your case github), to access that you need to identify yourself by the "secret" you are provided with along with your client ID, the "secret" must not be exposed in your client since it would give access to the user's private information to anyone.
This is a good explanation of the oauth process: https://docs.oracle.com/cd/E39820_01/doc.11121/gateway_docs/content/oauth_flows.html

So in short, if you want an MIT open source free library that can handle your authentication you need to have a server-side and you can use this library, if you want to use a good proprietary service that handles the authentication for you you need to use whatever auth0 provides. Anyway someone has to securely host your "oauth secrets" either you or a proprietary service.

There are other flows (like 2-legged oauth) which the client can request the authentication token directly but these flows are usually used for building a 3rd party client for the app you are getting your auth from (as in building a github client for github or a facebook client for facebook) and this library does not cover that scenario (usually each service has documentation for it or provide an sdk).

**Comment by nigel-dewar** [comment]
@KyRooZz . can you publish a copy of your source on github and I'll have a look at it.

cheers

Nige.

**Comment by KyRooZz** [comment]
Hi,

@ronzeidman Thank you for your help, now I understand a little bit better the data flow needed. I saw Auth0 but considering you have to pay... I'm not very interested and I think it can be very limited.

Basically, my application is going to receive multiple providers, I am going to loop through the JSON object to display the buttons and what I want is that the user can connect himself via the providers, so I want just to get his email address, and compared it from my side with my database and say "if your email address exists, you can log in in our application". So what I want is just to get his information and specially his email address.

@nigel-dewar  Here is a link to the project, it is really a test project that I downloaded, and I just tried to use ng2-ui-auth in order to get users information. There is no server side or anything else, just a simple button, which open a popup and that's it... I quickly checked the example with the server but I don't really get what happen after saying "yes authorize application..." via the client auth.authenticate function. How is it linked ? (Sorry I'm a beginner with Angular) :)

Link : https://github.com/KyRooZz/Angular2---Authenticate

**Comment by KyRooZz** [comment]
Hello guys,

I managed to run the ng2-ui-auth example so now I know how it works between the client/server and why the server-side is needed (Exchange authorization code for access token...) :)

But I have few more questions (sorry to be annoying with this "server-side") : 

- I checked satellizer librairy and they say that it is possible to make a client-side using `set responseType to 'token'`, is it possible with ng2-ui-auth ?

- I saw that the responseType is only available with Oauth2 providers, so that means only Twitter is not going to work ?

- If it works with ng2-ui-auth, what is going to happen after saying "authorize the application" ? Because now it's going directly to the server and exchanging authorization code for access token etc...

Thanks again for the lib !

**Comment by ronzeidman** [comment]
:laughing: It's actually here in my code: https://github.com/ronzeidman/ng2-ui-auth/blob/master/src/oauth2.ts#L67
Maybe I was a bit drunk copying that from the satellizer library, anyway in that flow I don't exchange the authorization token in the server so I guess the server should return the actual "token" instead of the "authorization 'code'". Try it and tell us if it works.

**Comment by ronzeidman** [comment]
Does it work?

**Comment by KyRooZz** [comment]
Ahh yes sorry, I managed to get the the token directly without any server side following an example that I found with Satellizer and with what you said in your previous comment.

Thanks again ! :)
